### PR TITLE
add range variant set

### PR DESCRIPTION
**Pull request by Kaitaan**
add RangeVariantSet to allow manually setting bucketing ranges.

This is a recreation of https://github.com/reddit/baseplate/pull/230

👓 @mattknox @spladug

### CODE REVIEW

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from .base import VariantSet
+
+
+class RangeVariantSet(VariantSet):
+    """ Variant Set designed to take fixed bucket ranges.
```
Comment (spladug) [inline baseplate/experiments/variant_sets/range_variant_set.py]: nit: please follow pep257 guidelines for docstrings. the first line should be a standalone summary and then there should be a blank followed by as much extra docs as you need.

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    This VariantSet allows manually setting bucketing ranges.
+    It takes in a variant name, then the range of buckets in
+    that should be assigned to that variant. This enables user-defined
+    bucketing algorithms, as well as simplifies the ability to adjust
+    range sizes in special circumstances.
+    """
+
+    def __init__(self, variants, num_buckets=1000):
+        """ :param list variants: Array of dicts, each containing the keys 'name',
```
Comment (spladug) [inline baseplate/experiments/variant_sets/range_variant_set.py]: i don't think you can docstring __init__ here can you? does this show up in the docs when building locally? those params would probably go in the class docstring.

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+            if variant.get('range_start') is None or variant.get('range_end') is None:
+                raise ValueError('Variant range invalid: {}'.format(self.variants))
+            range_size = variant.get('range_end') - variant.get('range_start')
+            total_size += int(range_size * self.num_buckets)
+
+        if total_size > self.num_buckets:
+            raise ValueError('Sum of all variants is greater than 100%')
+
+    def choose_variant(self, bucket):
+        """Deterministically choose a variant. Every call with the same bucket
```
Comment (spladug) [inline baseplate/experiments/variant_sets/range_variant_set.py]: nit: same nit about pep257. the summary line should fit on one line.

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+        if self.variants is None:
+            raise ValueError('No variants provided')
+
+        if len(self.variants) < 1:
+            raise ValueError("RangeVariant experiments expect at least one variant")
+
+        total_size = 0
+        for variant in self.variants:
+            if variant.get('range_start') is None or variant.get('range_end') is None:
```
Comment (spladug) [inline baseplate/experiments/variant_sets/range_variant_set.py]: it'd be more pythonic to do:

```python
try:
    range_size = variant["range_end"] - variant["range_start"]
except KeyError:
    raise ValueError("Variant range invalid: {}".format(self.variants))

total_size += int(range_size * self.num_buckets)
```

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    def __contains__(self, item):
+        for variant in self.variants:
+            if variant.get('name') == item:
+                return True
+
+        return False
+
+    def _validate_variants(self):
+
+        if self.variants is None:
```
Comment (spladug) [inline baseplate/experiments/variant_sets/range_variant_set.py]: you could condense these two checks into 

```python
if not self.variants:
    raise ValueError("No variants provided")
```

File: baseplate/experiments/variant_sets/range_variant_set.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
+from .base import VariantSet
```
[the remark this answers is not in the document]
Comment (Kaitaan) [reply baseplate/experiments/variant_sets/range_variant_set.py]: Can do. Should I add it to all the other files that don't have it too? (all the other files in this folder, everything in the targeting folder, etc)?

**Comment by Kaitaan** [comment]
Finally had some time to clean this up a bit.

**Pull request merged**